slider/beatmap.py

```diff
@@ -1414,31 +1415,150 @@ def ar(self,
 
         return ar
 
-    @lazyval
-    def hit_objects_no_spinners(self):
-        """The hit objects with spinners filtered out.
+    @memoize
+    def hit_objects(self,
+                    spinners=True,
+                    stacking=True,
+                    easy=False,
+                    hard_rock=False,
+                    double_time=False,
+                    half_time=False):
+        """Retrieve hit_objects.
+
+        Parameters
+        ----------
+        spinners : bool, optional
+            If spinners should be included.
+        stacking : bool, optional
+            If stacking should be calculated.
+        easy : bool, optional
+            Calculate stacking with the easy mod enabled.
+        hard_rock : bool, optional
+            Hit_objects will be flipped on the y axis. Calculate stacking with the hard rock mod enabled.
+        half_time : bool, optional
+            Calculate stacking with the half time mod enabled.
+        double_time : bool, optional
+            Calculate stacking with the double time mod enabled.
+
+        Returns
+        -------
+        hit_objects : list[HitObject]
+            The effective AR value.
         """
-        return tuple(e for e in self.hit_objects if not isinstance(e, Spinner))
+
+        hit_objects = self._hit_objects
+
+        if hard_rock:
+            for hit_object in hit_objects:
+                p = hit_object.position
+                p_new = Position(p.x, 384 - p.y)
+                hit_object.position = p_new
```

Comment:
just a note:
> ```python
>    elif easy:
>        hit_objects = [ob.easy for ob in hit_objects]
Easy don't really affect the hit_objects, as in they neither change time or position so this doesn't make sense.

> This logic should really live on the slider object itself since it know how hard rock affects itself.

It already does, no idea why I didn't just use that.

<sub><sup>also you messed up the half_time branch</sub></sup>

Reply:
Yeah, my bad. I really want that to be symmetric because almost everything else that cares about a hit object depends on ez/hr and ht/dt

> also you messed up the half_time branch

code review saves lives